lib/livebook/settings.ex

```diff
@@ -117,4 +120,160 @@ defmodule Livebook.Settings do
+
+  @doc """
+  Updates an environment variable from given changeset.
+
+  With success, notifies interested processes about environment variables
+  data change. Otherwise, it will return an error tuple with changeset.
+  """
+  @spec update_env_var(EnvironmentVariable.t(), map()) ::
+          {:ok, EnvironmentVariable.t()} | {:error, Ecto.Changeset.t()}
+  def update_env_var(%EnvironmentVariable{} = env_var, attrs) do
+    env_var
+    |> EnvironmentVariable.changeset(attrs)
+    |> apply_action(:update)
+    |> update_env_var()
+  end
+
+  defp update_env_var({:ok, %EnvironmentVariable{} = env_var}) do
+    if env_var_exists?(env_var.key) do
+      save_env_var(env_var)
+    else
+      {:error,
+       env_var
+       |> change_env_var()
+       |> add_error(:key, "does not exists")}
+    end
+  end
+
+  defp update_env_var({:error, changeset}), do: {:error, %{changeset | action: :validate}}
+
+  defp save_env_var(env_var) do
+    attributes = env_var |> Map.from_struct() |> Map.to_list()
+
+    with :ok <- storage().insert(:environment_variables, env_var.id, attributes),
+         :ok <- broadcast_environment_variables_change() do
+      {:ok, env_var}
+    end
+  end
+
+  @doc """
+  Deletes an environment variable from given id.
+
+  Also, it notifies interested processes about environment variables data change.
+  """
+  @spec delete_env_var(String.t()) :: :ok
+  def delete_env_var(id) do
+    storage().delete(:environment_variables, id)
+    broadcast_environment_variables_change()
+  end
+
+  def clean_env_vars do
+    for env_var <- fetch_env_vars(), do: delete_env_var(env_var.id)
+    :ok
+  end
+
+  @doc """
+  Returns an `%Ecto.Changeset{}` for tracking environment variable changes.
+  """
+  @spec change_env_var(EnvironmentVariable.t(), map()) :: Ecto.Changeset.t()
+  def change_env_var(%EnvironmentVariable{} = env_var, attrs \\ %{}) do
+    env_var
+    |> EnvironmentVariable.changeset(attrs)
+    |> Map.put(:action, :validate)
+  end
+
+  @doc """
+  Subscribes to updates in environment variables information.
+
+  ## Messages
+
+    * `{:environment_variables_changed, environment_variables}`
+
+  """
+  @spec subscribe() :: :ok | {:error, term()}
+  def subscribe do
+    Phoenix.PubSub.subscribe(Livebook.PubSub, "environment_variables")
+  end
+
+  @doc """
+  Unsubscribes from `subscribe/0`.
+  """
+  @spec unsubscribe() :: :ok
+  def unsubscribe do
+    Phoenix.PubSub.unsubscribe(Livebook.PubSub, "environment_variables")
+  end
+
+  @doc """
+  Notifies interested processes about environment variables data change.
+
+  Broadcasts `{:environment_variables_changed, environment_variables}` message under the `"environment_variables"` topic.
+  """
+  @spec broadcast_environment_variables_change() :: :ok
+  def broadcast_environment_variables_change do
+    Phoenix.PubSub.broadcast(
+      Livebook.PubSub,
+      "environment_variables",
+      {:environment_variables_changed, fetch_env_vars()}
+    )
+  end
```

Comment:
There are a lot of new functions here, so let's move them to `Livebook.Settings.EnvVar` (I suggest using EnvVar for conciseness, but `EnvironmentVariable` is also fine).

Reply:
Actually, let's keep it in this module, but we can trim it down. Some ideas:

* Make `broadcast_environment_variables_change` private.
* Make `subscribe` and `unsubscribe` to happen on the `settings` topic so it is more generic.

But I would still suggest renaming the schema to `Livebook.Settings.EnvVar`.